ApplicationLibCode/FileInterface/RifRoffFileTools.cpp

```diff
@@ -0,0 +1,479 @@
+    // i,j), transform it according to zoffset and zscale, and put it in to
+    // zcornsv in reverse order.
+
+    // As i and j order and size is the same for both zcornsv and zdata, we can
+    // ignore it here and place one pillar at the time regardless of how many
+    // pillars there are.
+
+    size_t nzcorn    = zcornsv.size();
+    size_t nsplitenz = splitenz.size();
+    size_t nzdata    = zdata.size();
+
+    size_t num_row = 4 * static_cast<size_t>( nz );
+    if ( nzcorn % num_row != 0 ) throw std::runtime_error( "Incorrect size of zcorn." );
+    if ( nsplitenz != nzcorn / 4 ) throw std::runtime_error( "Incorrect size of splitenz." );
+
+    std::vector<float> pillar( num_row, 0.0 );
+    size_t             it_zdata = 0, it_splitenz = 0, it_zcorn = 0;
+    while ( it_zcorn < nzcorn )
+    {
+        for ( size_t it_pillar = 0; it_pillar < num_row; )
+        {
+            char split = splitenz[it_splitenz++];
+            if ( split == 1 )
+            {
+                // There is one value for this corner which
+                // we must duplicate 4 times in zcornsv
+                if ( it_zdata >= nzdata ) throw std::runtime_error( "Incorrect size of zdata" );
+                float val = ( zdata[it_zdata++] + zoffset ) * zscale;
+                for ( int n = 0; n < 4; n++ )
+                {
+                    pillar[it_pillar++] = val;
+                }
+            }
+            else if ( split == 4 )
+            {
+                // There are four value for this corner which
+                // we must duplicate 4 times in zcornsv
+                if ( it_zdata + 3 >= nzdata ) throw std::runtime_error( "Incorrect size of zdata" );
+                // As we place the pillar in reverse order into zcornsv,
+                // we must put zdata in reverse order into pillar to
+                // preserve n,s,w,e directions.
+                pillar[it_pillar + 3] = ( zdata[it_zdata++] + zoffset ) * zscale;
+                pillar[it_pillar + 2] = ( zdata[it_zdata++] + zoffset ) * zscale;
+                pillar[it_pillar + 1] = ( zdata[it_zdata++] + zoffset ) * zscale;
+                pillar[it_pillar]     = ( zdata[it_zdata++] + zoffset ) * zscale;
+                it_pillar += 4;
+            }
+            else
+            {
+                throw std::runtime_error( "Unsupported split type" );
+            }
+        }
+        // Put the pillar into zcornsv in reverse order
+        for ( size_t it_pillar = num_row; it_pillar >= 1; )
+        {
+            zcornsv[it_zcorn++] = pillar[--it_pillar];
+        }
+    }
+
+    if ( it_splitenz != nsplitenz ) throw std::runtime_error( "Incorrect size of splitenz." );
+    if ( it_zdata != nzdata ) throw std::runtime_error( "Incorrect size of zdata" );
+    if ( it_zcorn != nzcorn ) throw std::runtime_error( "Incorrect size of zcorn." );
+}
+
+//--------------------------------------------------------------------------------------------------
+///
```

Comment:
Maybe a small figure describing the numbering of nodes? See top of `RifReaderEclipseOutput.cpp`